### PR TITLE
Fix division by zero warnings from cuda forall

### PR DESCRIPTION
**Pull request by MrBurmark**
# Fix warnings from nvcc
Nvcc warnes about divisions by zero that are there in some template instantiations but are never run due to conditionals.
Partially specialize the forall dimension calculator classes to avoid this.

- This PR is a bugfix
- It does the following :
  - Fixes a warning I noticed in RAJAPerf

### CODE REVIEW

File: include/RAJA/policy/cuda/forall.hpp [hunk 1]
```diff
@@ -78,50 +78,78 @@ struct ForallDimensionCalculator<::RAJA::iteration_mapping::Direct,
                                  ::RAJA::cuda::IndexGlobal<dim, BLOCK_SIZE, GRID_SIZE>,
                                  UniqueMarker>
 {
-  static_assert(BLOCK_SIZE > 0 || BLOCK_SIZE == named_usage::unspecified, "block size may not be ignored with forall");
-  static_assert(GRID_SIZE > 0 || GRID_SIZE == named_usage::unspecified, "grid size may not be ignored with forall");
+  static_assert(BLOCK_SIZE > 0, "block size may not be ignored with forall");
```
Comment (rhornung67) [inline include/RAJA/policy/cuda/forall.hpp:81]:
```suggestion
  static_assert(BLOCK_SIZE > 0, "block size > 0 must be specified with forall");
```
Is this more clear? Similarly, with the following static assert.

File: include/RAJA/policy/cuda/forall.hpp [hunk 2]
```diff
@@ -130,63 +158,91 @@ struct ForallDimensionCalculator<::RAJA::iteration_mapping::StridedLoop,
...
```
Comment (rhornung67) [inline include/RAJA/policy/cuda/forall.hpp:161]: Same comment as above to clarify the comment.

File: include/RAJA/policy/cuda/forall.hpp [hunk 3]
```diff
@@ -130,63 +158,91 @@ struct ForallDimensionCalculator<::RAJA::iteration_mapping::StridedLoop,
...
+  }
+};
 
-      } else {
+template<named_dim dim, int BLOCK_SIZE, typename UniqueMarker>
+struct ForallDimensionCalculator<::RAJA::iteration_mapping::StridedLoop,
+                                 ::RAJA::cuda::IndexGlobal<dim, BLOCK_SIZE, named_usage::unspecified>,
+                                 UniqueMarker>
+{
+  static_assert(BLOCK_SIZE > 0, "block size may not be ignored with forall");
```
Comment (rhornung67) [inline include/RAJA/policy/cuda/forall.hpp:205]: Ditto.

File: include/RAJA/policy/hip/forall.hpp [hunk 4]
```diff
@@ -79,50 +79,78 @@ struct ForallDimensionCalculator<::RAJA::iteration_mapping::Direct,
                                  ::RAJA::hip::IndexGlobal<dim, BLOCK_SIZE, GRID_SIZE>,
                                  UniqueMarker>
 {
-  static_assert(BLOCK_SIZE > 0 || BLOCK_SIZE == named_usage::unspecified, "block size may not be ignored with forall");
-  static_assert(GRID_SIZE > 0 || GRID_SIZE == named_usage::unspecified, "grid size may not be ignored with forall");
+  static_assert(BLOCK_SIZE > 0, "block size may not be ignored with forall");
```
Comment (rhornung67) [inline include/RAJA/policy/hip/forall.hpp:82]: Ditto again.

File: include/RAJA/policy/hip/forall.hpp [hunk 5]
```diff
@@ -79,50 +79,78 @@ struct ForallDimensionCalculator<::RAJA::iteration_mapping::Direct,
...
+};
 
-      internal::set_hip_dim<dim>(dims.threads, static_cast<IdxT>(max_sizes.first));
-      internal::set_hip_dim<dim>(dims.blocks, RAJA_DIVIDE_CEILING_INT(len, static_cast<IdxT>(max_sizes.first)));
+template<named_dim dim, int GRID_SIZE, typename UniqueMarker>
+struct ForallDimensionCalculator<::RAJA::iteration_mapping::Direct,
+                                 ::RAJA::hip::IndexGlobal<dim, named_usage::unspecified, GRID_SIZE>,
+                                 UniqueMarker>
+{
+  static_assert(GRID_SIZE > 0, "grid size may not be ignored with forall");
```
Comment (rhornung67) [inline include/RAJA/policy/hip/forall.hpp:106]: same

File: include/RAJA/policy/hip/forall.hpp [hunk 6]
```diff
@@ -79,50 +79,78 @@ struct ForallDimensionCalculator<::RAJA::iteration_mapping::Direct,
...
+};
 
-      internal::set_hip_dim<dim>(dims.threads, static_cast<IdxT>(IndexGetter::block_size));
-      internal::set_hip_dim<dim>(dims.blocks, RAJA_DIVIDE_CEILING_INT(len, static_cast<IdxT>(IndexGetter::block_size)));
+template<named_dim dim, int BLOCK_SIZE, typename UniqueMarker>
+struct ForallDimensionCalculator<::RAJA::iteration_mapping::Direct,
+                                 ::RAJA::hip::IndexGlobal<dim, BLOCK_SIZE, named_usage::unspecified>,
+                                 UniqueMarker>
+{
+  static_assert(BLOCK_SIZE > 0, "block size may not be ignored with forall");
```
Comment (rhornung67) [inline include/RAJA/policy/hip/forall.hpp:125]: same

File: include/RAJA/policy/hip/forall.hpp [hunk 7]
```diff
@@ -131,63 +159,91 @@ struct ForallDimensionCalculator<::RAJA::iteration_mapping::StridedLoop,
                                  ::RAJA::hip::IndexGlobal<dim, BLOCK_SIZE, GRID_SIZE>,
                                  UniqueMarker>
 {
-  static_assert(BLOCK_SIZE > 0 || BLOCK_SIZE == named_usage::unspecified, "block size may not be ignored with forall");
-  static_assert(GRID_SIZE > 0 || GRID_SIZE == named_usage::unspecified, "grid size may not be ignored with forall");
+  static_assert(BLOCK_SIZE > 0, "block size may not be ignored with forall");
```
Comment (rhornung67) [inline include/RAJA/policy/hip/forall.hpp:162]: ditto. same, etc.

File: include/RAJA/policy/hip/forall.hpp [hunk 8]
```diff
@@ -131,63 +159,91 @@ struct ForallDimensionCalculator<::RAJA::iteration_mapping::StridedLoop,
...
+  }
+};
 
-      } else {
+template<named_dim dim, int BLOCK_SIZE, typename UniqueMarker>
+struct ForallDimensionCalculator<::RAJA::iteration_mapping::StridedLoop,
+                                 ::RAJA::hip::IndexGlobal<dim, BLOCK_SIZE, named_usage::unspecified>,
+                                 UniqueMarker>
+{
+  static_assert(BLOCK_SIZE > 0, "block size may not be ignored with forall");
```
Comment (rhornung67) [inline include/RAJA/policy/hip/forall.hpp:206]: same

File: include/RAJA/policy/cuda/forall.hpp (hunk 1 above)
Comment (MrBurmark) [reply include/RAJA/policy/cuda/forall.hpp:81]: These comments are a bit cryptic now that you point it out. There are currently four possible states, positive, 0, -1 and "other." Positive means a compile time size. 0 means unspecified and is handled in another partial specialization. -1 means ignored and is not allowed in forall as you must use blocks and threads. "Other" doesn't have a meaning and is always invalid. This static assert is mainly guarding against -1/ignored but also guards against "other" values.

Comment (MrBurmark) [reply include/RAJA/policy/cuda/forall.hpp:81]: So positive and 0 values are valid in forall. I was trying to call out that ignored is not a valid value here.

**Review by rhornung67 — APPROVED** [review]
Much better. Thank you.

**Pull request merged**